EventGenerator/src/RPCGun_module.cc

```diff
@@ -56,214 +54,207 @@ namespace mu2e {
+    , pionCaptureSpectrum_{&randomFlat_,&randomUnitSphere_}
+  {
+    produces<mu2e::StageParticleCollection>();
+    produces<mu2e::EventWeight>();
+    if(RPCType_ == "ExternalRPC") process_ = ProcessCode::ExternalRPC;
+    else if(RPCType_ == "InternalRPC") process_ = ProcessCode::InternalRPC;
+    else {
+      throw   cet::exception("BADINPUT")
+        <<"RPCGun::produce(): no such process, must be InternalRPC or ExternalRPC";
+    } 
+    if ( doHistograms_ ) {
         art::ServiceHandle<art::TFileService> tfs;
         art::TFileDirectory tfdir = tfs->mkdir( "RPCGun" );
 
         _hmomentum     = tfdir.make<TH1F>( "hmomentum", "Produced photon momentum", 100,  40.,  140.  );
-
-        if(generateInternalConversion_){
+        _time        = tfdir.make<TH1F>("time"       , "time", 20000,0,1);
+        if(RPCType_ == "InternalRPC"){
           _hElecMom  = tfdir.make<TH1F>("hElecMom" , "Produced electron momentum", 140,  0. , 140.);
           _hPosiMom  = tfdir.make<TH1F>("hPosiMom" , "Produced positron momentum", 140,  0. , 140.);
           _hMee      = tfdir.make<TH1F>("hMee"     , "M(e+e-) "           , 200,0.,200.);
           _hMeeVsE   = tfdir.make<TH2F>("hMeeVsE"  , "M(e+e-) vs E"       , 200,0.,200.,200,0,200);
           _hMeeOverE = tfdir.make<TH1F>("hMeeOverE", "M(e+e-)/E "         , 200, 0.,1);
-          _hy        = tfdir.make<TH1F>("hy"       , "y = (ee-ep)/|pe+pp|", 200,-1.,1.);
+          _hy        = tfdir.make<TH1F>("hy"       , "y = (ee-ep)/|pe+pp|", 200,-1.,1.);      
         }
       }
-
-    }
-
-  //================================================================
-
-  void RPCGun::beginRun(art::Run& run) {
-    protonPulse_.reset( new ProtonPulseRandPDF( eng_, protonPset_ ) );
   }
-
-  void RPCGun::produce(art::Event& event) {
-
-    std::unique_ptr<GenParticleCollection> output(new GenParticleCollection);
-    std::unique_ptr<FixedTimeMap> timemap(new FixedTimeMap);
-
-    ConditionsHandle<AcceleratorParams> accPar("ignored");
-    double _mbtime = accPar->deBuncherPeriod;
-
-    IO::StoppedParticleTauNormF stop;
-    if (tmin_ > 0){
-      while (true){
-        const auto& tstop = stops_.fire();
-        timemap->SetTime(protonPulse_->fire());
-        if (tstop.t+timemap->time() < 0 || tstop.t+timemap->time() > tmin_){
-          if (applySurvivalProbability_){
-            double weight = exp(-tstop.tauNormalized)*survivalProbScaling_;
-            if (weight > 1)
-              std::cout << "WEIGHT TOO HIGH " << weight << " " << fmod(tstop.t,_mbtime) << std::endl;
-            double rand = randomFlat_.fire();
-            if (weight > rand){
-              stop = tstop;
-              break;
+  
+  double RPCGun::MakeEventWeight(art::Event& event){ 
+    const auto simh = event.getValidHandle<SimParticleCollection>(simsToken_); 
+    double weight = 0.;
+    double tau = 0.;
+    const PhysicsParams& gc = *GlobalConstantsHandle<PhysicsParams>();
+    for(const auto& p : *simh) {
```

Comment:
This is only computing the weight for the primary particle.  And it is incrementing the proper time also for non-pions.  I don't see how this works, but maybe I'm missing something.  Physically, since you have suppressed only pion lifetime, you want the weight to depend only on the proper time of pions, and ONLY on pions in the decay chain of the stopped pion (additional pions produced by the proton are irrelevant).  You also need to be careful of the split due to beam resampling, which splits the SimParticle for the pion which enters the DS in 2.  In that case, you want to add the proper times before and after the split before computing the weight.

Reply:
OK, so there should be some additonal code inside the mu2ePrimary loop which selects pions and I think we want mu2eKillerVolume as the tag we need to add, is that correct?